### PR TITLE
Alovelltroy/add file back end

### PR DESCRIPTION
**Pull request by alexlovelltroy**
This pull request introduces a new secure storage system that implements the SecureStorage interface for managing secrets locally using AES-GCM encryption with a master key derived via HKDF. It includes the implementation of encryption and decryption functions, a `LocalStore` for managing secrets, and comprehensive testing for all functionalities. Additionally, it updates dependencies to support the new features.

### Core Features and Implementation:

* **Encryption and Decryption Functions**:
  - Added `deriveAESKey`, `encryptAESGCM`, and `decryptAESGCM` functions in `encryption.go` to provide AES-GCM encryption with keys derived from a master key using HKDF.

* **Local Secret Store**:
  - Implemented `LocalStore` in `localStoreAdapter.go` to handle secure storage of secrets in a JSON file, with methods for storing, retrieving, listing, and deleting secrets. It ensures encryption using AES-GCM and supports automatic file reloads when modified externally.

### Example Usage

There is an example implementation in the examples/LocalStore directory that can be used to store and retrieve secrets by name.  It requires a 32 bit master key which can be generated by the example cli as well.  Set the master key to `MASTER_KEY` before attempting to store or retrieve any secrets.

### Testing and Validation:

* **Unit Tests for Encryption**:
  - Added tests in `encryption_test.go` to validate key derivation, encryption, and decryption functionalities, ensuring correctness and consistency.

* **Unit Tests for LocalStore**:
  - Added comprehensive tests in `localStoreAdapter_test.go` to verify `LocalStore` methods, including edge cases like empty keys, nil values, external file modifications, and custom struct handling.

### Dependency Updates:

* **Remove vendor directory**
* **`go.mod` Updates**:
  - Updated `go` version and added `golang.org/x/crypto` as a new dependency to support cryptographic operations. [[1]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L3-R3) [[2]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R24)

### CODE REVIEW

File: localStoreAdapter.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	if secrets == nil {
+		secrets, err = loadSecrets(filename)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load secrets from file: %v", err)
+		}
+	}
+
+	lastModTime, _ := getModTime(filename) // changed code (ignoring error here for brevity)
```
Comment (shunr-hpe) [inline localStoreAdapter.go:175]:
```suggestion
	lastModTime, lastSize, _ := getFileStats(filename) // file changed code (ignoring error here for brevity)
```

File: localStoreAdapter.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+	return &LocalStore{
+		masterKey:   masterKey,
+		filename:    filename,
+		Secrets:     secrets,
+		lastModTime: lastModTime, // changed code
```
Comment (shunr-hpe) [inline localStoreAdapter.go:181]:
```suggestion
		lastModTime: lastModTime, // file changed code
		lastSize:    lastSize,
```

File: localStoreAdapter.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	defer file.Close()
+
+	store := make(map[string]string)
+	decoder := json.NewDecoder(file)
+	err = decoder.Decode(&store)
+	return store, err
+}
+
+func getModTime(filename string) (time.Time, error) {
```
Comment (shunr-hpe) [inline localStoreAdapter.go:232]:
```suggestion
func getFileStats(filename string) (t time.Time, size int64, err error) {
```

File: localStoreAdapter.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+	info, err := os.Stat(filename)
+	if err != nil {
+		return time.Time{}, err
```
Comment (shunr-hpe) [inline localStoreAdapter.go:235]:
```suggestion
		return time.Time{}, -1, err
```

File: localStoreAdapter.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	return info.ModTime(), nil
```
Comment (shunr-hpe) [inline localStoreAdapter.go:237]:
```suggestion
	return info.ModTime(), info.Size(), nil
```

File: localStoreAdapter.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+// reloadIfChanged reloads secrets from disk if the file has been modified
+func (l *LocalStore) reloadIfChanged() error {
+	currentModTime, err := getModTime(l.filename)
```
Comment (shunr-hpe) [inline localStoreAdapter.go:242]:
```suggestion
	currentModTime, currentSize, err := getFileStats(l.filename)
```

File: localStoreAdapter.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+	if err != nil {
+		return err
+	}
+	if currentModTime.After(l.lastModTime) {
```
Comment (shunr-hpe) [inline localStoreAdapter.go:246]:
```suggestion
	if currentModTime.After(l.lastModTime) || currentSize != l.lastSize {
```

File: localStoreAdapter_test.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+		}
+
+		// Simulate external change - note the updated JSON, changing the structure to a plain string
+		os.WriteFile(tmpFile, []byte(`{"reloadLookupKey":"changedExternally"}`), 0644)
+
+		// Lookup should detect external change and reload before reading.
+		// We'll use an empty interface to reflect the new type in the file (a string).
+		var output interface{}
+		if err := store.Lookup("reloadLookupKey", &output); err == nil {
+			t.Fatalf("Failed to lookup after external modification: %v", err)
```
Comment (shunr-hpe) [inline localStoreAdapter_test.go:227]:
```suggestion
			t.Fatalf("Failed to fail while looking up an unencrypted key after an out of band write of the file: %v", err)
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,15 @@ All notable changes to this project will be documented in this file.
...
 
+## [1.18.0] - 2025-04-30
+
+### Security
+
+- Added a file based storage implementation
+
+
+
+
```
Comment (shunr-hpe) [inline CHANGELOG.md:16]:
```suggestion

```

**Review by mbuchmann-hpe — APPROVED** [review]

**Review by shunr-hpe — APPROVED** [review]
